File: practice/binary_tree/binary_tree_path_sum.cc
// sum.cc --- 
// 
// Filename: sum.cc
// Description: 
// 
// Author:    Yu Lu
// Email:     yulu@utexas.edu
// Github:    https://github.com/SuperYuLu 
// 
// Created: Thu Nov 22 17:23:03 2018 (-0600)
// Version: 
// Last-Updated: Thu Nov 22 17:23:11 2018 (-0600)
//           By: yulu
//     Update #: 1
// 

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
  bool hasPathSum(TreeNode* root, int sum) {
    if(root)
      return checkPathSum(root, sum);
    else
      return false;
  }
    
private:
  bool checkPathSum(TreeNode* root, int sum)
  {   bool res = false;

    if (root -> left)
      res = res || checkPathSum(root -> left, sum - (root -> val));
    if (root -> right)
      res = res || checkPathSum(root -> right, sum  - (root -> val));

    if (!root -> left && !root -> right && root -> val == sum)
      res = true;
    return res;
        
  }
};

File: practice/binary_tree/binary_tree_path_sum.py
# binary_tree_path_sum.py --- 
# 
# Filename: binary_tree_path_sum.py
# Description: 
# 
# Author:    Yu Lu
# Email:     yulu@utexas.edu
# Github:    https://github.com/SuperYuLu 
# 
# Created: Sat Nov  3 14:40:20 2018 (-0500)
# Version: 
# Last-Updated: Sat Nov  3 14:46:41 2018 (-0500)
#           By: yulu
#     Update #: 3
# 




class Solution:
    '''
    Definition for a binary tree node.
    class TreeNode:
        def __init__(self, x):
            self.val = x
            self.left = None
            self.right = None
    '''
    # Bottom up recursive solution
    # 60 ms, beat 46%
    def hasPathSum(self, root, sumVal = None):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """     
        res = False 
        
        if root:       
            sumVal -= root.val
            if not root.left and not root.right and sumVal == 0:
                return True          
            return self.hasPathSum(root.left, sumVal) or self.hasPathSum(root.right, sumVal)
    
        else:
            return False 
